### PR TITLE
ticket-675: Double check that User Schedule times are related to the user’s time zone

### PR DESCRIPTION
**Pull request by danloa**
Fixes [http://tickets.pm4overflow.com/tickets/675](http://tickets.pm4overflow.com/tickets/675)

- Improves the 'date' extension for FormalExpressions to be able to use a timezone. 
- If no time zone is provided to this new version of the 'date' function, the configured timezone in .env is used (so it will work as before)
- Unit tests were added
- The process used to determine the assignment has been modified to use this new 'date' function passing the user's timezone.

### CODE REVIEW

**Comment by pmkerberos** [comment]
QA server was successfully deployed https://bugfix-ticket-675-processmaker.ci.processmaker.net

**Comment by pmkerberos** [comment]
QA server was successfully deployed https://bugfix-ticket-675-processmaker.ci.processmaker.net

**Comment by pmkerberos** [comment]
QA server was successfully deployed https://bugfix-ticket-675-processmaker.ci.processmaker.net

**Comment by pmkerberos** [comment]
QA server was successfully deployed https://bugfix-ticket-675-processmaker.ci.processmaker.net

File: database/processes/AssignmentProcess.bpmn [hunk 1]
```diff
@@ -43,14 +43,14 @@
...
+    <bpmn:sequenceFlow id="node_14" name="SCHEDULED" sourceRef="node_11" targetRef="node_13" pm:config="{&#34;update_data&#34;:{&#34;variable&#34;:&#34;schedule&#34;,&#34;expression&#34;:&#34;get(user(user_id).schedule, date(\&#34;D\&#34;, null, user(user_id).timezone))&#34;}}">
       <bpmn:conditionExpression xsi:type="bpmn:tFormalExpression">user(user_id).status=="SCHEDULED"</bpmn:conditionExpression>
     </bpmn:sequenceFlow>
     <bpmn:endEvent id="node_15" name="Assign to user_id">
       <bpmn:incoming>node_16</bpmn:incoming>
     </bpmn:endEvent>
     <bpmn:sequenceFlow id="node_16" name="IN SCHEDULE" sourceRef="node_13" targetRef="node_15" pm:config="{&#34;update_data&#34;:{&#34;variable&#34;:&#34;assign_to&#34;,&#34;expression&#34;:&#34;user_id&#34;}}">
-      <bpmn:conditionExpression xsi:type="bpmn:tFormalExpression">schedule.active and ((schedule.from &lt; schedule.to and date("H:i") &gt;= schedule.from and date("H:i") &lt; schedule.to) or (schedule.from &gt; schedule.to and (date("H:i") &gt;= schedule.from or date("H:i") &lt; schedule.to)))</bpmn:conditionExpression>
```
Comment (caleeli) [inline database/processes/AssignmentProcess.bpmn:53]: These changes are perfect, please could you add these changes to the package too

File: database/seeds/AssignmentProcess.php [hunk 2]
```diff
@@ -16,30 +16,38 @@ class AssignmentProcess extends Seeder
      */
     public function run()
     {
-        if ($this->processExists()) {
-            return;
-        }
+//        if ($this->processExists()) {
```
Comment (caleeli) [inline database/seeds/AssignmentProcess.php:19]: Please remove this commented code

File: tests/unit/ProcessMaker/Models/FormalExpressionTest.php [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        $this->assertEquals($currentTimeInMinutes->format('Y-m-d H:i'), $feelDateInMinutes);
+
+        // Validate that timestamp for function date works
+        $timestamp = time();
+        $formalExp->setBody("date('Y-m-d H:i:s', $timestamp, 'Africa/Djibouti')");
+        $feelDate = $formalExp([]);
+        $date = new DateTime("now", new \DateTimeZone('Africa/Djibouti'));
+        $date->setTimestamp($timestamp);
+        $this->assertEquals($date->format('Y-m-d H:i:s'), $feelDate);
+    }
```
Comment (caleeli) [inline tests/unit/ProcessMaker/Models/FormalExpressionTest.php:53]: Please add one additional test to check the condition used in the process works as expected:
Something like:
```
// Set the current date and time in UTC
Carbon::setTestNow(Carbon::parse('2021-08-31 12:00'))->timezone('UTC'); 
// Time config to test
$user_tz = "America/La_Paz";
$user_schedule_time = "08:00"; // am
// Check the expression date('H:i', null, user_tz) == user_schedule_time

```

**Review by caleeli — APPROVED** [review]

**Comment by pmkerberos** [comment]
QA server was successfully deployed https://bugfix-ticket-675-processmaker.ci.processmaker.net

**Comment by pmkerberos** [comment]
QA server was successfully deployed https://bugfix-ticket-675-processmaker.ci.processmaker.net

**Comment by pmkerberos** [comment]
QA server was successfully deployed https://develop-processmaker.ci.processmaker.net

**Comment by pmkerberos** [comment]
QA server was successfully deployed https://bugfix-ticket-675-processmaker.ci.processmaker.net

**Comment by pmkerberos** [comment]
QA server was successfully deployed https://bugfix-ticket-675-processmaker.ci.processmaker.net

**Pull request merged**